### PR TITLE
Generalize read/to_json to work with pathlib.Path and raw json text

### PR DESCRIPTION
**Pull request by Strilanc**
Fixes https://github.com/quantumlib/Cirq/issues/2422

Example serialize:

```python
cirq.to_json(cirq.X)
```

```
'{\n  "cirq_type": "_PauliX",\n  "exponent": 1.0,\n  "global_shift": 0.0\n}'
```

Example deserialize:

```python
cirq.read_json(json_text='{\n  "cirq_type": "_PauliX",\n  "exponent": 1.0,\n  "global_shift": 0.0\n}')
```

```
cirq.X
```

### CODE REVIEW

File: cirq/protocols/json.py [hunk 1]
```diff
@@ -312,21 +344,39 @@ def to_json(obj: Any, file_or_fn, *, indent=2, cls=CirqEncoder):
             party classes, prefer adding the _json_dict_ magic method
             to your classes rather than overriding this default.
     """
-    if isinstance(file_or_fn, str):
+    if file_or_fn is None:
+        buffer = io.StringIO()
+        to_json(obj, buffer)
+        buffer.seek(0)
+        return buffer.read()
```
Comment (mpharrigan) [inline cirq/protocols/json.py]: If you want, you can replace this block with
```python
json.dumps(obj, indent=indent, cls=cls)
```

File: cirq/protocols/json.py [hunk 2]
```diff
@@ -335,6 +385,14 @@ def read_json(
             to indicate that it cannot resolve the given cirq_type and that
             the next resolver should be tried.
     """
+    if (file_or_fn is None) == (json_text is None):
+        raise ValueError('Must specify ONE of "file_or_fn" or "json".')
+
+    if json_text is not None:
+        file_or_fn = io.StringIO()
```
Comment (mpharrigan) [inline cirq/protocols/json.py]: You can replace this with 
```python
json.loads(json_text, object_hook=obj_hook)
```

Comment (Strilanc) [reply cirq/protocols/json.py]: Done.

File: cirq/protocols/json.py (hunk 1 above)
Comment (Strilanc) [reply cirq/protocols/json.py]: Done.

**Pull request merged**